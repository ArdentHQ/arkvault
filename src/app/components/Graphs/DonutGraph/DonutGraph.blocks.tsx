import React, { useEffect, useRef } from "react";

import { ContentInsideCircleProperties } from "./DonutGraph.contracts";

export const ContentInsideCircle: React.VFC<ContentInsideCircleProperties> = ({ renderFunction, size }) => {
	const reference = useRef<HTMLDivElement | null>(null);

	useEffect(() => {
		const container = reference.current;

		if (!container) {
			return;
		}

		const { width, height } = container.getBoundingClientRect();

		// Center content vertically and horizontally.
		container.style.left = `${Math.floor(size / 2 - width / 2)}px`;
		container.style.top = `${Math.floor(size / 2 - height / 2)}px`;
	}, [size]);

	if (!renderFunction) {
		return <></>;
	}

	return (
		<div className="absolute" ref={reference}>
			{renderFunction()}
		</div>
	);
};
