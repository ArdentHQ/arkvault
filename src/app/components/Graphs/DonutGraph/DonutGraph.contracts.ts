import React from "react";
import { GraphAnimation } from "@/app/components/Graphs/GraphHoverAnimation/GraphHoverAnimation.contract";
import { GraphDataPoint } from "@/app/components/Graphs/Graphs.contracts";

const BACKGROUND_CIRCLE_SPACING = 16;
const GRAPH_MARGIN = 32;
const RADIUS_HOVER_INCREMENT = 16;
const SEGMENT_SPACING = 20;

interface DonutGraphProperties {
	data: GraphDataPoint[];
	size: number;
	renderTooltip?: (dataPoint: GraphDataPoint) => JSX.Element;
	renderContentInsideCircle?: () => JSX.Element;
}

interface DonutGraphConfig {
	circleCommonProperties: React.SVGProps<SVGCircleElement>;
	circumference: number;
	circumferenceHover: number;
	radius: number;
	radiusHover: number;
}

interface DonutGraphCircle {
	circleProperties: React.SVGProps<SVGCircleElement>;
	animations: GraphAnimation[];
}

interface ContentInsideCircleProperties {
	renderFunction: (() => JSX.Element) | undefined;
	size: number;
}

type UseDonutGraphHook = (
	data: GraphDataPoint[],
	size: number,
) => {
	backgroundCircle: React.SVGProps<SVGCircleElement>;
	circles: DonutGraphCircle[];
};

export { BACKGROUND_CIRCLE_SPACING, GRAPH_MARGIN, RADIUS_HOVER_INCREMENT, SEGMENT_SPACING };

export type {
	ContentInsideCircleProperties,
	DonutGraphCircle,
	DonutGraphConfig,
	DonutGraphProperties,
	UseDonutGraphHook,
};
