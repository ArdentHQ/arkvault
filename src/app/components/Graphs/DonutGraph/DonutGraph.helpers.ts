import React, { useMemo } from "react";

import {
	BACKGROUND_CIRCLE_SPACING,
	DonutGraphCircle,
	DonutGraphConfig,
	GRAPH_MARGIN,
	RADIUS_HOVER_INCREMENT,
	SEGMENT_SPACING,
	UseDonutGraphHook,
} from "./DonutGraph.contracts";

const useDonutGraph: UseDonutGraphHook = (data, size) => {
	const { circleCommonProperties, circumference, circumferenceHover, radius, radiusHover } =
		useMemo<DonutGraphConfig>(() => {
			const diameter = size - GRAPH_MARGIN * 2;
			const center = size / 2;

			const radius = diameter / 2;
			const radiusHover = radius + RADIUS_HOVER_INCREMENT;

			return {
				circleCommonProperties: {
					cx: center,
					cy: center,
					fill: "transparent",
					r: radius,
					strokeLinecap: "round",
					strokeWidth: 16,
				},
				circumference: diameter * Math.PI,
				circumferenceHover: radiusHover * 2 * Math.PI,
				radius,
				radiusHover,
			};
		}, [size]);

	const backgroundCircle = useMemo<React.SVGProps<SVGCircleElement>>(
		() => ({
			...circleCommonProperties,
			className: "fill-current text-theme-secondary-200 dark:text-black",
			fill: "inherit",
			r: radius - BACKGROUND_CIRCLE_SPACING,
			strokeWidth: 0,
		}),
		[size], // eslint-disable-line react-hooks/exhaustive-deps
	);

	const circles = useMemo<DonutGraphCircle[]>(() => {
		const items: DonutGraphCircle[] = [];

		// When there is only 1 segment avoid showing
		// the space between start and end of it.
		const spacing = data.length > 1 ? SEGMENT_SPACING : 0;

		// By default circle dasharray starts at 90Â°, which would be 25 given a
		// circumference of 100. Set initial offset to 25 (in proportion to the
		// circumference) to make sure that the first slice will start at 12:00.
		let strokeDashoffset = circumference * 0.25;
		let strokeDashoffsetHover = circumferenceHover * 0.25;

		for (const { color, value } of data) {
			const dash = (circumference * value) / 100;
			const dashHover = (circumferenceHover * value) / 100;

			const gap = (circumference * (100 - value)) / 100;
			const gapHover = (circumferenceHover * (100 - value)) / 100;

			// stroke-dashoffset has to be equal to the sum of the
			// lengths of the previous segments or they will overlap.
			strokeDashoffset += dash;
			strokeDashoffsetHover += dashHover;

			const strokeDasharray = [dash - spacing, gap + spacing].join(" ");
			const strokeDasharrayHover = [dashHover - spacing, gapHover + spacing].join(" ");

			items.push({
				animations: [
					{ attribute: "r", from: radius, to: radiusHover },
					{ attribute: "stroke-dasharray", from: strokeDasharray, to: strokeDasharrayHover },
					{ attribute: "stroke-dashoffset", from: strokeDashoffset, to: strokeDashoffsetHover },
				],
				circleProperties: {
					...circleCommonProperties,
					className: `stroke-current text-theme-${color}`,
					strokeDasharray,
					strokeDashoffset,
				},
			});
		}

		return items;
	}, [circumference, data]); // eslint-disable-line react-hooks/exhaustive-deps

	return {
		backgroundCircle,
		circles,
	};
};

export { useDonutGraph };
