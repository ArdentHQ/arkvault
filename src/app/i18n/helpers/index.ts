import { translations as CONTACTS } from "@/domains/contact/i18n";
import { translations as DASHBOARD } from "@/domains/dashboard/i18n";
import { translations as ERROR } from "@/domains/error/i18n";
import { translations as EXCHANGE } from "@/domains/exchange/i18n";
import { translations as MESSAGE } from "@/domains/message/i18n";
import { translations as PROFILE } from "@/domains/profile/i18n";
import { translations as SETTINGS } from "@/domains/setting/i18n";
import { translations as TRANSACTION } from "@/domains/transaction/i18n";
import { translations as VOTE } from "@/domains/vote/i18n";
import { translations as WALLETS } from "@/domains/wallet/i18n";
import { translations as COMMON } from "@/app/i18n/common/i18n";

export const buildTranslations = () => ({
	COMMON,
	CONTACTS,
	DASHBOARD,
	ERROR,
	EXCHANGE,
	MESSAGE,
	PROFILE,
	SETTINGS,
	TRANSACTION,
	VOTE,
	WALLETS,
});
