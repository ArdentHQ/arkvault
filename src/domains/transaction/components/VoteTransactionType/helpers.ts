import { DTO } from "@ardenthq/sdk";
import { DTO as ProfilesDTO } from "@ardenthq/sdk-profiles";

/**
 * @TODO: This can be handled in SDK to retrieve voting and unvoting public keys
 * through a common method/format, whether it's a signed or confirmed transaction.
 * Currently votes() unvotes() methods don't exist signed transaction data
 *  and `data` is a function in signed transaction and an object in confirmed transaction
 */
export const extractVotingData = ({ transaction }: { transaction: DTO.RawTransactionData }) => {
	if (transaction.isConfirmed()) {
		return {
			unvotes: transaction.unvotes(),
			votes: transaction.votes(),
		};
	}

	return signedTransactionVotingData({ transaction });
};

const signedTransactionVotingData = ({ transaction }: { transaction: ProfilesDTO.ExtendedSignedTransactionData }) => {
	const data = transaction.data().data();
	const votes: string[] = [];
	const unvotes: string[] = [];

	// array of publicKeys
	const rawVotes: string[] = data?.asset?.votes ?? [];

	for (const publicKey of rawVotes) {
		const destination = publicKey.startsWith("-") ? unvotes : votes;
		destination.push(publicKey.replace(/^[+-]+/, ""));
	}

	return {
		unvotes,
		votes,
	};
};
