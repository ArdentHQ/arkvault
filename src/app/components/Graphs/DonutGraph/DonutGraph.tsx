import { sortBy } from "@payvo/sdk-helpers";
import React, { useMemo } from "react";

import { ContentInsideCircle } from "./DonutGraph.blocks";
import { DonutGraphProperties } from "./DonutGraph.contracts";
import { useDonutGraph } from "./DonutGraph.helpers";
import { useGraphTooltip } from "@/app/components/Graphs/Graphs.shared";
import { GraphHoverAnimation } from "@/app/components/Graphs/GraphHoverAnimation";

export const DonutGraph: React.VFC<DonutGraphProperties> = ({
	data,
	size,
	renderTooltip,
	renderContentInsideCircle,
}) => {
	const sortedData = useMemo(() => sortBy(data, (item) => item.value), [data]);

	const { Tooltip, getMouseEventProperties } = useGraphTooltip(renderTooltip, "donut");
	const { circles, backgroundCircle } = useDonutGraph(sortedData, size);

	const renderCircles = () =>
		circles.map(({ circleProperties, animations }, index) => (
			<g key={index} data-testid="DonutGraph__item">
				<circle
					{...circleProperties}
					id={`circleTrackLine__${index}`}
					data-testid="DonutGraph__item-track-line"
					className="stroke-current text-theme-secondary-300 dark:text-theme-secondary-800"
					strokeWidth={2}
					pointerEvents="none"
				/>
				<circle
					{...circleProperties}
					id={`circleHoverArea__${index}`}
					data-testid="DonutGraph__item-hover-area"
					strokeWidth={40}
					opacity={0}
					pointerEvents="visibleStroke"
					{...getMouseEventProperties(sortedData[index])}
				/>
				<circle {...circleProperties} pointerEvents="none">
					<GraphHoverAnimation targetElementId={`circleHoverArea__${index}`} animations={animations} />
				</circle>
			</g>
		));

	return (
		<div className="relative">
			<Tooltip />

			<ContentInsideCircle renderFunction={renderContentInsideCircle} size={size} />

			<svg width={size} height={size} data-testid="DonutGraph__svg">
				<circle {...backgroundCircle} />

				{renderCircles()}
			</svg>
		</div>
	);
};
