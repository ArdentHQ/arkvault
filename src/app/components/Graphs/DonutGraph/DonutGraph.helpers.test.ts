import { renderHook } from "@testing-library/react-hooks";

import { useDonutGraph } from "./DonutGraph.helpers";

describe("DonutGraph.helpers", () => {
	describe("useDonutGraph", () => {
		it.each([
			{
				data: [
					{ color: "success-600", data: {}, value: 50 },
					{ color: "warning-600", data: {}, value: 30 },
					{ color: "info-600", data: {}, value: 20 },
				],
			},
			{
				data: [{ color: "success-600", data: {}, value: 100 }],
			},
		])("should plot data to svg circles needed to render the donut", ({ data }) => {
			const { result } = renderHook(() => useDonutGraph(data, 100));

			expect(result.current).toMatchSnapshot();
		});
	});
});
