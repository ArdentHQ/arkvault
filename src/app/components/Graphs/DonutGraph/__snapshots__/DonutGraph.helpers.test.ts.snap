// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DonutGraph.helpers useDonutGraph should plot data to svg circles needed to render the donut 1`] = `
Object {
  "backgroundCircle": Object {
    "className": "fill-current text-theme-secondary-200 dark:text-black",
    "cx": 50,
    "cy": 50,
    "fill": "inherit",
    "r": 2,
    "strokeLinecap": "round",
    "strokeWidth": 0,
  },
  "circles": Array [
    Object {
      "animations": Array [
        Object {
          "attribute": "r",
          "from": 18,
          "to": 34,
        },
        Object {
          "attribute": "stroke-dasharray",
          "from": "36.548667764616276 76.54866776461628",
          "to": "86.81415022205297 126.81415022205297",
        },
        Object {
          "attribute": "stroke-dashoffset",
          "from": 84.82300164692441,
          "to": 160.22122533307947,
        },
      ],
      "circleProperties": Object {
        "className": "stroke-current text-theme-success-600",
        "cx": 50,
        "cy": 50,
        "fill": "transparent",
        "r": 18,
        "strokeDasharray": "36.548667764616276 76.54866776461628",
        "strokeDashoffset": 84.82300164692441,
        "strokeLinecap": "round",
        "strokeWidth": 16,
      },
    },
    Object {
      "animations": Array [
        Object {
          "attribute": "r",
          "from": 18,
          "to": 34,
        },
        Object {
          "attribute": "stroke-dasharray",
          "from": "13.929200658769766 99.1681348704628",
          "to": "44.08849013323179 169.53981031087417",
        },
        Object {
          "attribute": "stroke-dashoffset",
          "from": 118.75220230569417,
          "to": 224.30971546631127,
        },
      ],
      "circleProperties": Object {
        "className": "stroke-current text-theme-warning-600",
        "cx": 50,
        "cy": 50,
        "fill": "transparent",
        "r": 18,
        "strokeDasharray": "13.929200658769766 99.1681348704628",
        "strokeDashoffset": 118.75220230569417,
        "strokeLinecap": "round",
        "strokeWidth": 16,
      },
    },
    Object {
      "animations": Array [
        Object {
          "attribute": "r",
          "from": 18,
          "to": 34,
        },
        Object {
          "attribute": "stroke-dasharray",
          "from": "2.6194671058465104 110.47786842338604",
          "to": "22.725660088821186 190.90264035528475",
        },
        Object {
          "attribute": "stroke-dashoffset",
          "from": 141.37166941154067,
          "to": 267.0353755551325,
        },
      ],
      "circleProperties": Object {
        "className": "stroke-current text-theme-info-600",
        "cx": 50,
        "cy": 50,
        "fill": "transparent",
        "r": 18,
        "strokeDasharray": "2.6194671058465104 110.47786842338604",
        "strokeDashoffset": 141.37166941154067,
        "strokeLinecap": "round",
        "strokeWidth": 16,
      },
    },
  ],
}
`;

exports[`DonutGraph.helpers useDonutGraph should plot data to svg circles needed to render the donut 2`] = `
Object {
  "backgroundCircle": Object {
    "className": "fill-current text-theme-secondary-200 dark:text-black",
    "cx": 50,
    "cy": 50,
    "fill": "inherit",
    "r": 2,
    "strokeLinecap": "round",
    "strokeWidth": 0,
  },
  "circles": Array [
    Object {
      "animations": Array [
        Object {
          "attribute": "r",
          "from": 18,
          "to": 34,
        },
        Object {
          "attribute": "stroke-dasharray",
          "from": "113.09733552923255 0",
          "to": "213.62830044410595 0",
        },
        Object {
          "attribute": "stroke-dashoffset",
          "from": 141.3716694115407,
          "to": 267.0353755551324,
        },
      ],
      "circleProperties": Object {
        "className": "stroke-current text-theme-success-600",
        "cx": 50,
        "cy": 50,
        "fill": "transparent",
        "r": 18,
        "strokeDasharray": "113.09733552923255 0",
        "strokeDashoffset": 141.3716694115407,
        "strokeLinecap": "round",
        "strokeWidth": 16,
      },
    },
  ],
}
`;
